File: packages/contacts/src/__tests__/ContactsModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { ContactsModule } from '../ContactsModule.js';

describe('ContactsModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let contacts: ContactsModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    contacts = new ContactsModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  describe('getContacts', () => {
    it('sends get contacts request', async () => {
      const promise = contacts.getContacts({ query: 'John', limit: 10, offset: 0 });
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('contacts');
      expect(sent.action).toBe('getContacts');

      respondSuccess({ contacts: [{ id: '1', givenName: 'John', familyName: 'Doe', phoneNumbers: [], emailAddresses: [] }] });
      const result = await promise;
      expect(result.contacts).toHaveLength(1);
      expect(result.contacts[0].givenName).toBe('John');
    });
  });

  describe('getContact', () => {
    it('returns a single contact', async () => {
      const promise = contacts.getContact({ id: '1' });
      respondSuccess({ id: '1', givenName: 'Jane', familyName: 'Doe', phoneNumbers: [], emailAddresses: [] });
      const result = await promise;
      expect(result.givenName).toBe('Jane');
    });
  });

  describe('createContact', () => {
    it('creates a contact and returns id', async () => {
      const promise = contacts.createContact({ givenName: 'New', familyName: 'Contact' });
      respondSuccess({ id: 'new-1' });
      const result = await promise;
      expect(result.id).toBe('new-1');
    });
  });

  describe('deleteContact', () => {
    it('deletes a contact', async () => {
      const promise = contacts.deleteContact({ id: '1' });
      respondSuccess();
      await promise;
    });
  });

  describe('requestPermission', () => {
    it('requests permission', async () => {
      const promise = contacts.requestPermission();
      respondSuccess({ granted: true });
      const result = await promise;
      expect(result.granted).toBe(true);
    });
  });

  describe('getPermissionStatus', () => {
    it('returns permission status', async () => {
      const promise = contacts.getPermissionStatus();
      respondSuccess({ status: 'granted' });
      const result = await promise;
      expect(result.status).toBe('granted');
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors', async () => {
      const promise = contacts.getContacts();
      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Permission denied' },
      };
      transport.simulateIncoming(JSON.stringify(response));
      await expect(promise).rejects.toThrow('Permission denied');
    });
  });
});

File: android/contacts/src/test/kotlin/io/rynbridge/contacts/ContactsModuleTest.kt
package io.rynbridge.contacts

import io.rynbridge.core.*
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class ContactsModuleTest {

    @Test
    fun `getContacts returns contact list`() = runTest {
        val provider = MockContactsProvider()
        val module = ContactsModule(provider)
        val handler = module.actions["getContacts"]!!

        val result = handler(mapOf(
            "query" to BridgeValue.string("John"),
            "limit" to BridgeValue.int(10),
            "offset" to BridgeValue.int(0)
        ))
        val contacts = result["contacts"]?.arrayValue
        assertNotNull(contacts)
        assertEquals(1, contacts!!.size)
        assertEquals("John", contacts[0].dictionaryValue?.get("givenName")?.stringValue)
    }

    @Test
    fun `getContact returns single contact`() = runTest {
        val provider = MockContactsProvider()
        val module = ContactsModule(provider)
        val handler = module.actions["getContact"]!!

        val result = handler(mapOf("id" to BridgeValue.string("c1")))
        assertEquals("c1", result["id"]?.stringValue)
        assertEquals("John", result["givenName"]?.stringValue)
        assertEquals("Doe", result["familyName"]?.stringValue)
    }

    @Test
    fun `createContact returns id`() = runTest {
        val provider = MockContactsProvider()
        val module = ContactsModule(provider)
        val handler = module.actions["createContact"]!!

        val result = handler(mapOf(
            "givenName" to BridgeValue.string("Jane"),
            "familyName" to BridgeValue.string("Smith"),
            "phoneNumbers" to BridgeValue.array(listOf(
                BridgeValue.dict(mapOf(
                    "label" to BridgeValue.string("mobile"),
                    "number" to BridgeValue.string("555-1234")
                ))
            )),
            "emailAddresses" to BridgeValue.array(listOf(
                BridgeValue.dict(mapOf(
                    "label" to BridgeValue.string("work"),
                    "address" to BridgeValue.string("jane@example.com")
                ))
            ))
        ))
        assertEquals("new-id", result["id"]?.stringValue)
        assertEquals("Jane", provider.lastCreateGivenName)
    }

    @Test
    fun `updateContact calls provider`() = runTest {
        val provider = MockContactsProvider()
        val module = ContactsModule(provider)
        val handler = module.actions["updateContact"]!!

        val result = handler(mapOf(
            "id" to BridgeValue.string("c1"),
            "givenName" to BridgeValue.string("Johnny")
        ))
        assertTrue(result.isEmpty())
        assertEquals("c1", provider.lastUpdateId)
        assertEquals("Johnny", provider.lastUpdateGivenName)
    }

    @Test
    fun `deleteContact calls provider`() = runTest {
        val provider = MockContactsProvider()
        val module = ContactsModule(provider)
        val handler = module.actions["deleteContact"]!!

        val result = handler(mapOf("id" to BridgeValue.string("c1")))
        assertTrue(result.isEmpty())
        assertEquals("c1", provider.lastDeleteId)
    }

    @Test
    fun `pickContact returns contact`() = runTest {
        val provider = MockContactsProvider()
        val module = ContactsModule(provider)
        val handler = module.actions["pickContact"]!!

        val result = handler(emptyMap())
        val contact = result["contact"]?.dictionaryValue
        assertNotNull(contact)
        assertEquals("c1", contact!!["id"]?.stringValue)
    }

    @Test
    fun `requestPermission returns granted`() = runTest {
        val provider = MockContactsProvider()
        val module = ContactsModule(provider)
        val handler = module.actions["requestPermission"]!!

        val result = handler(emptyMap())
        assertEquals(true, result["granted"]?.boolValue)
    }

    @Test
    fun `getPermissionStatus returns status`() = runTest {
        val provider = MockContactsProvider()
        val module = ContactsModule(provider)
        val handler = module.actions["getPermissionStatus"]!!

        val result = handler(emptyMap())
        assertEquals("granted", result["status"]?.stringValue)
    }

    @Test
    fun `module name and version`() {
        val provider = MockContactsProvider()
        val module = ContactsModule(provider)
        assertEquals("contacts", module.name)
        assertEquals("0.1.0", module.version)
    }
}

private class MockContactsProvider : ContactsProvider {
    var lastCreateGivenName: String? = null
    var lastUpdateId: String? = null
    var lastUpdateGivenName: String? = null
    var lastDeleteId: String? = null

    private val mockContact = ContactData(
        id = "c1",
        givenName = "John",
        familyName = "Doe",
        phoneNumbers = listOf(ContactPhoneData("mobile", "555-0000")),
        emailAddresses = listOf(ContactEmailData("home", "john@example.com"))
    )

    override suspend fun getContacts(query: String?, limit: Int, offset: Int): List<ContactData> =
        listOf(mockContact)

    override suspend fun getContact(id: String): ContactData = mockContact

    override suspend fun createContact(
        givenName: String,
        familyName: String,
        phoneNumbers: List<Pair<String, String>>,
        emailAddresses: List<Pair<String, String>>
    ): String {
        lastCreateGivenName = givenName
        return "new-id"
    }

    override suspend fun updateContact(
        id: String,
        givenName: String?,
        familyName: String?,
        phoneNumbers: List<Pair<String, String>>?,
        emailAddresses: List<Pair<String, String>>?
    ) {
        lastUpdateId = id
        lastUpdateGivenName = givenName
    }

    override suspend fun deleteContact(id: String) {
        lastDeleteId = id
    }

    override suspend fun pickContact(): ContactData = mockContact

    override suspend fun requestPermission(): Boolean = true

    override suspend fun getPermissionStatus(): String = "granted"
}
